### ISSUE TITLE
[no-unnecessary-condition] Potential side effects not taken into account

### ISSUE THREAD

**Issue by davidaurelio** [typescript-eslint/typescript-eslint#1459]
*no-unnecessary-condition* does not take potential side effects into account when assigning to a variable in a higher scope or a property, calling a function/method, and eventually checking that variable/property again. Instead, the value of the assignment is treated as constant. 

This assumption is only true for locals, though. Hence, the rule should not flag any non-locals. 

(There is, of course, the possibility of checking whether the function invoked comes from a higher scope than the variable in question, but implementing that logic seems over the top).

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition" : [ "error" ]
  }
}
```

```TS
type A = { value(): number };

// Repro with condition in higher scope + side effects
let big = false;

function checkA(a: A) {
  big = big || a.value() > 0xffff;
}

function check(as: Array<A>): void {
  big = false;
  for (const a of as) checkA(a);
  if (big) console.log("Found a value > 0xffff"); // errors
}


// Repro with class properties + side effects
class Check {
  big = false;

  checkA(a: A) {
    this.big = this.big || a.value() > 0xffff;
  }

  check(as: Array<A>) {
    this.big = false;
    for (const a of as) this.checkA(a);
    if (this.big) console.log("Found a value > 0xffff"); // errors
  }
}

```

**Expected Result**

No lint errors

**Actual Result**

`Unnecessary conditional, value is always falsy` errors for lines 12 and 25.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.16.0` |
| `@typescript-eslint/parser`        | `2.16.0` |
| `TypeScript`                       | `3.7.4` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `13.3.0` |
| `yarn`                              | `1.19.2` |

**Comment by davidaurelio** [comment]
Gist with full project for minimal repro: https://gist.github.com/10e385aa5d3817e8bbf0de87f94a6528

**Comment by bradzacher** [comment]
This isn't something we can control.
TypeScript doesn't widen the type of the variable, because it doesn't do control-flow analysis which includes called functions.

![image](https://user-images.githubusercontent.com/7462525/72450447-41058600-376f-11ea-9dd3-a0f0bbd208d4.png)
![image](https://user-images.githubusercontent.com/7462525/72450480-4e227500-376f-11ea-8adc-84febc7af04e.png)

In both cases TypeScript reports the type as `false`, not `boolean`.
We can only act on the information given by TypeScript.

I would consider annotating your variables with `: boolean` so TypeScript reports the type correctly.

**Comment by MikeyBurkman** [comment]
For anyone who comes across this like I did, it's essentially a flaw in how TS does its flow analysis. There's a massive discussion on it dating back to 2016 [here](https://github.com/microsoft/TypeScript/issues/9998) 

I'm not sure there actually is a good workaround, unfortunately. The suggested workaround of adding `: boolean` to the declaration won't matter, as TS is already inferring the type as `boolean` when it is declared. The issue comes from TS narrowing the type later on, but not knowing that a function has side effects that might make the type narrowing no longer accurate.
